account-compression/sdk/tests/accountCompression.test.ts

```diff
@@ -0,0 +1,684 @@
+      );
+    });
+
+    it("Random attacker fails to fake the existence of a leaf by autocompleting proof", async () => {
+      const maliciousLeafHash = crypto.randomBytes(32);
+      const maliciousLeafHash1 = crypto.randomBytes(32);
+      const nodeProof: Buffer[] = [];
+      for (let i = 0; i < DEPTH; i++) {
+        nodeProof.push(Buffer.alloc(32));
+      }
+
+      // Root - make this nonsense so it won't match what's in ChangeLog, thus forcing proof autocompletion
+      const replaceIx = createReplaceIx(
+        payer,
+        splCMTKeypair.publicKey,
+        Buffer.alloc(32),
+        maliciousLeafHash,
+        maliciousLeafHash1,
+        0,
+        nodeProof
+      );
+
+      try {
+        await execute(provider, [replaceIx], [payer]);
+        assert(
+          false,
+          "Attacker was able to succesfully write fake existence of a leaf"
+        );
+      } catch (e) { }
+
+      const splCMT = await getConcurrentMerkleTree(connection, splCMTKeypair.publicKey);
+
+      assert(
+        getCMTActiveIndex(splCMT) === 0,
+        "CMT updated its active index after attacker's transaction, when it shouldn't have done anything"
+      );
+    });
+    it("Random attacker fails to fake the existence of a leaf by autocompleting proof", async () => {
+      const maliciousLeafHash = crypto.randomBytes(32);
+      const maliciousLeafHash1 = crypto.randomBytes(32);
+      const nodeProof: Buffer[] = [];
+      for (let i = 0; i < DEPTH; i++) {
+        nodeProof.push(Buffer.alloc(32));
+      }
+
+      // Root - make this nonsense so it won't match what's in CL, and force proof autocompletion
+      const replaceIx = createReplaceIx(
+        payer,
+        splCMTKeypair.publicKey,
+        Buffer.alloc(32),
+        maliciousLeafHash,
+        maliciousLeafHash1,
+        0,
+        nodeProof
+      );
+
+      try {
+        await execute(provider, [replaceIx], [payer]);
+        assert(
+          false,
+          "Attacker was able to succesfully write fake existence of a leaf"
+        );
+      } catch (e) { }
+
+      const splCMT = await getConcurrentMerkleTree(provider.connection, splCMTKeypair.publicKey);
+
+      assert(
+        getCMTActiveIndex(splCMT) === 0,
+        "CMT updated its active index after attacker's transaction, when it shouldn't have done anything"
+      );
+    });
+  });
+  describe(`Canopy test`, () => {
+    const DEPTH = 5;
+    it("Testing canopy for appends and replaces on a full on chain tree", async () => {
+      [splCMTKeypair, offChainTree] = await createTreeOnChain(
+        payer,
+        0,
+        DEPTH,
+        8,
+        DEPTH // Store full tree on chain
+      );
+
+      let leaves: Array<number>[] = [];
+      let i = 0;
+      let stepSize = 4;
+      while (i < 2 ** DEPTH) {
+        let ixs: TransactionInstruction[] = [];
+        for (let j = 0; j < stepSize; ++j) {
+          const newLeaf = Array.from(Buffer.alloc(32, i + 1));
```

Comment:
To make the leaves unique, should this be?
```suggestion
          const newLeaf = Array.from(Buffer.alloc(32, i + j));
```

Reply:
Leaves are actually unique in this case, since `i` is being incremented by step size each iteration. 

But because this portion of the test is complicated, I added some docs.

This functionality is annoyingly nuanced, so I would prefer to rewrite this into several small test functions via unit tests (in a separate follow up PR)

Reply:
All right, I'll take your word for it then!